exchanges/poloniex/poloniex_websocket.go

```diff
@@ -87,196 +94,318 @@ func (p *Poloniex) WsHandleData() {
 		select {
 		case <-p.Websocket.ShutdownC:
 			return
-
 		default:
 			resp, err := p.WebsocketConn.ReadMessage()
 			if err != nil {
 				p.Websocket.ReadMessageErrors <- err
 				return
 			}
 			p.Websocket.TrafficAlert <- struct{}{}
-			var result interface{}
-			err = json.Unmarshal(resp.Raw, &result)
+			err = p.wsHandleData(resp.Raw)
 			if err != nil {
 				p.Websocket.DataHandler <- err
-				continue
 			}
-			switch data := result.(type) {
-			case map[string]interface{}:
-				// subscription error
-				p.Websocket.DataHandler <- errors.New(data["error"].(string))
-			case []interface{}:
-				chanID := int(data[0].(float64))
-				if len(data) == 2 && chanID != wsHeartbeat {
-					if checkSubscriptionSuccess(data) {
-						if p.Verbose {
-							log.Debugf(log.ExchangeSys,
-								"%s websocket subscribed to channel successfully. %d",
-								p.Name,
-								chanID)
+		}
+	}
+}
+
+func (p *Poloniex) wsHandleData(respRaw []byte) error {
+	var result interface{}
+	err := json.Unmarshal(respRaw, &result)
+	if err != nil {
+		return err
+	}
+	if data, ok := result.([]interface{}); ok {
+		if len(data) == 0 {
+			return errors.New("c'mon man, no data")
+		}
+		if len(data) == 1 {
+			// heartbeat
+			return nil
+		}
+		if len(data) == 2 {
+			// subscription acknowledgement / heartbeat
+			return nil
+		}
+		if channelID, ok := data[0].(float64); ok {
+			switch channelID {
+			case ws24HourExchangeVolumeID:
+				return nil
+			case wsAccountNotificationID:
+				if notificationsArray, ok := data[2].([]interface{}); ok {
+					if _, ok := notificationsArray[0].([]interface{}); ok {
+						for i := 0; i < len(notificationsArray); i++ {
+							if notification, ok := (notificationsArray[i]).([]interface{}); ok {
+								switch notification[0].(string) {
+								case "o":
+									var amount float64
+									amount, err = strconv.ParseFloat(notification[2].(string), 64)
+									if err != nil {
+										return err
+									}
+									var oStatus order.Status
+									var oType = notification[3].(string)
+									switch {
+									case amount > 0 && (oType == "f" || oType == "s"):
+										oStatus = order.PartiallyFilled
+									case amount == 0 && (oType == "f" || oType == "s"):
+										oStatus = order.Filled
+									case amount > 0 && oType == "c":
+										oStatus = order.PartiallyCancelled
+									case amount == 0 && oType == "c":
+										oStatus = order.Cancelled
+									}
+									response := &order.Modify{
+										RemainingAmount: amount,
+										Exchange:        p.Name,
+										ID:              strconv.FormatFloat(notification[1].(float64), 'f', -1, 64),
+										Type:            order.Limit,
+										Status:          oStatus,
+										AssetType:       asset.Spot,
+									}
+									p.Websocket.DataHandler <- response
+								case "n":
+									var timeParse time.Time
+									timeParse, err = time.Parse("2006-01-02 15:04:05", notification[6].(string))
```

Comment:
Since this layout is used quite a lot, can make this a const similar to how we've done it in other changes for easy usage by other devs